### PR TITLE
Support attribute context update.

### PR DESCRIPTION
**Pull request by qiwzhang**
Also moved the Attributes struct related code into its own files.

### CODE REVIEW

File: src/attribute.cc [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (geeknoid) [inline src/attribute.cc]: You should change all the copyrights in the repo to say Copyright Istio Authors instead.

Comment (qiwzhang) [inline src/attribute.cc]: will do in another PR.  Thanks.
